{"cells": [{"cell_type": "markdown", "id": "determined-funds", "metadata": {}, "source": ["# RNA-seq data analysis"]}, {"cell_type": "markdown", "id": "victorian-airfare", "metadata": {}, "source": ["**First Summer School on Bioinformatics**  \n", "June - 2021  \n", "Tampere University  \n", "Author: Ebrahim Afyounian  "]}, {"cell_type": "markdown", "id": "great-sleeping", "metadata": {}, "source": ["In this exercise, we are going to work with **gene expression quantification** from **RNA-seq** experiment from [this](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE70466) study.\n", "\n", "- &#x1F4DD; **Activity**: Follow the link above to familiarize yourself with the type of samples in the data set.\n", "    - What are the two cell lines from which this data set is composed of?\n", "    - Which cell line in this experiment represents the normal cells?\n", "    - Which cell line in this experiment represents the tumor cells?\n", "    - How many samples are there from each cell line? \n", "    - Which sequencing platform was used for sequencing?"]}, {"cell_type": "markdown", "id": "cordless-maryland", "metadata": {}, "source": ["&#x26A0; **Note**: In what follows, the word \"cell\" refers to a box in this notebook that can be run by pressing \"<kbd>ctrl</kbd>+<kbd>Enter</kbd>\" buttons. If we want to refer to a biological cell, we will make it clear."]}, {"cell_type": "code", "execution_count": null, "id": "corresponding-newman", "metadata": {}, "outputs": [], "source": ["## This is a cell; run it by pressing \"ctrl+Enter\"\n", "1+1"]}, {"cell_type": "markdown", "id": "varying-allergy", "metadata": {}, "source": ["## Data preparation"]}, {"cell_type": "markdown", "id": "commercial-soldier", "metadata": {}, "source": ["Here, we jump over the details about how the abundances of each transcript in each sample has been determined for this experiment. If you are interested however, follow [this](https://pachterlab.github.io/kallisto/about) link.\n", "\n", "The result of quantification is available in the following file:\n", "`https://raw.githubusercontent.com/eafyounian/2021_summer_school_bioinf/main/resources/GSE70466-expression.txt`"]}, {"cell_type": "markdown", "id": "southeast-minimum", "metadata": {}, "source": ["To be able to analyze the data using R programming language, we need to load it in R environment.\n", "\n", "In R, we can load a file with different functions such as `read.table()` or `read.csv()`. \n", "\n", "To learn how to use a particular function in R, we can read the help/manual page for that function. For example, to read the manual for `read.table()` function, we can use `help(read.table)` or `?read.table`.\n", "- &#x1F4DD; **Activity**: Try running `?read.table` in the following cell and spend a few moment to familiarize yourself with the function."]}, {"cell_type": "code", "execution_count": null, "id": "reliable-athens", "metadata": {}, "outputs": [], "source": ["?read.table"]}, {"cell_type": "markdown", "id": "radical-gospel", "metadata": {}, "source": ["Running the following cell, load the quantification into a data frame called `df`.\n"]}, {"cell_type": "code", "execution_count": null, "id": "understood-glucose", "metadata": {}, "outputs": [], "source": ["df = read.table('https://raw.githubusercontent.com/eafyounian/2021_summer_school_bioinf/main/resources/GSE70466-expression.txt',\n", "                sep='\\t', \n", "                header=1, \n", "                stringsAsFactors = FALSE,\n", "                row.names=1,\n", "               )"]}, {"cell_type": "markdown", "id": "particular-affect", "metadata": {}, "source": ["We can use the `head()` function, like `head(df)` to inspect the first few rows in the data frame. \n", "- &#x1F4DD; **Activity**: Based on the output of the following cell, tell what the columns and the rows are (i.e. whether they are genes or samples)?"]}, {"cell_type": "code", "execution_count": null, "id": "overall-covering", "metadata": {}, "outputs": [], "source": ["head(df)"]}, {"cell_type": "markdown", "id": "declared-montana", "metadata": {}, "source": ["To see how many rows and how many column are there in our data frame `df`, we can use the `dim()` function that returns the _**dim**entions_ of the data frame.\n", "- &#x1F4DD; **Activity**: Based on the output of the following cell, how many genes have been tried to be quantified?"]}, {"cell_type": "code", "execution_count": null, "id": "imported-equilibrium", "metadata": {}, "outputs": [], "source": ["## Do the activity in this cell"]}, {"cell_type": "markdown", "id": "painful-garbage", "metadata": {}, "source": ["Even though it seems that there are `35238` rows in the data, not all of them will be informative because there are some rows that have similar values across all the columns. This could be a row with all `0`s or another number. We can discard these rows (this is an instance of **data preparation**). \n", "\n", "Let's see how many of such rows do exist in our data frame. One way to find out is to find the variance across  all the columns for each row and check the results. If the result for a row is `0`, we know that all columns  for that row have equal values (for example try this and check the result: `var(c(1, 1, 1))`).\n", "\n", "\n", "To calculate the variance over the columns for each row, we could write a `for` loop that calculates the variance one rows at a time, but we can also use R `apply()` function `apply(df, 1, var)` for this purpose. This function, applies the `var()` function to each as row denoted by `1` as the second function argument (see `?apply` for more info). \n", "\n", "\n", "Let's run the following cell and look at the first `20` rows. We can see that there are `2` rows with variance equal to `0`."]}, {"cell_type": "code", "execution_count": null, "id": "administrative-cement", "metadata": {}, "outputs": [], "source": ["df_vars = apply(df, 1, var)\n", "head(df_vars, 20)"]}, {"cell_type": "markdown", "id": "fluid-queensland", "metadata": {}, "source": [" "]}, {"cell_type": "markdown", "id": "expressed-oasis", "metadata": {}, "source": [" "]}, {"cell_type": "markdown", "id": "suburban-kennedy", "metadata": {}, "source": ["Now, let's remove the rows that have variance `0`. The other way to formulate this task is to **keep** the rows that do **not** have variance `0`. We can achieve this by running the following cell.\n", "\n", "Let's see what is happening in the following cell: `df_vars != 0` results in a Boolean vector where the value for a particular row is equal to `TRUE` if the the var is **non-zero** and `FALSE` otherwise. Next, we use this vector to create a subset of our data frame `df` and assign the results to a variable called `df_nonzero_var`. The `df[df_vars != 0, ]` tells R to keep only rows the are `TRUE` after the `df_vars != 0` expression is evaluated and also to keep all the columns.\n", "\n", "- &#x1F4DD; **Activity**: After running the following cell, check the dimensions of the new variable i.e. `df_nonzero_var` and tell how many of the rows were **filtered out**."]}, {"cell_type": "code", "execution_count": null, "id": "occupied-rebound", "metadata": {}, "outputs": [], "source": ["df_nonzero_var = df[df_vars != 0,]"]}, {"cell_type": "code", "execution_count": null, "id": "collect-metro", "metadata": {}, "outputs": [], "source": ["## Do the activity in this cell"]}, {"cell_type": "markdown", "id": "continental-berlin", "metadata": {}, "source": [" "]}, {"cell_type": "markdown", "id": "severe-showcase", "metadata": {}, "source": [" "]}, {"cell_type": "markdown", "id": "widespread-appraisal", "metadata": {}, "source": ["So far, we have loaded the data and performed some data preparation on it. Now, let's do some analysis. Each analysis may require more data preparation but we discuss them when needed."]}, {"cell_type": "markdown", "id": "cultural-brooklyn", "metadata": {}, "source": ["## Principal component analysis (PCA)"]}, {"cell_type": "markdown", "id": "designing-special", "metadata": {}, "source": ["Often, we are interested to visually inspect our data and check whether we can find any noticeable pattern in it. For example, we might be interested to see how similar or different they are from each other. \n", "\n", "When doing visualizations, we can go up to 3 dimensions. However, we have noticed that our data has thousands of rows and several columns (i.e. our data is high-dimensional). Thus we need a way to reduce the dimensions in our data set to be able to visualize it and yet this reduction in the dimensions should preserve the information in the data set as much as possible. \n", "\n", "One way to achieve this goal is to use a statistical method called **Principal Component Analysis** (PCA). In our particular application, this method takes gene expression values and transforms them to what is called as **principal components** (PCs). These are a set of linearly uncorrelated features that capture the sources of variance in the data.\n", "\n", "Once we have the PCs, we can use the first two or three principal components and visualize them in either a 2-dimensional (2-D) or 3-dimensional (3-D) scatter plot respectively.\n", "\n", "After visualization, if two data points (i.e. two RNA-seq samples) are close together, it means that they are highly correlated or in other words they are similar in their gene expression profiles.\n", "\n", "One thing that we need to make sure before PCA analysis, is to check whether our data is **centered** around `0` and whether our data is **scaled** to unit variance (i.e. `1`) because to work properly, PCA requires these conditions to be satisfied. If not, we need to tell the R function that does the PCA analysis for us to center and scale the data.\n", "\n", "&#x1F4DD; **Activity**: Use `rowMeans()` function to see whether the data is centered. Also inspect the `df_vars` to check whether the data is scaled to unit variance."]}, {"cell_type": "code", "execution_count": null, "id": "little-housing", "metadata": {}, "outputs": [], "source": ["## Do the activity in this cell"]}, {"cell_type": "markdown", "id": "unavailable-artist", "metadata": {}, "source": ["Now, let's perform PCA analysis on our data. We can use  `prcomp()` function to do so (run `?prcomp` to familiarize yourself with the function). \n", "\n", "This function accepts two arguments `scale` and `center` that we can set based on the result of the activity above. They tell the function whether we want to scale or center the data.\n", "\n", "**Note**: `prcomp()` function expects the rows to be samples and columns to be genes. But our data is in the opposite order. Thus we need to transpose our data. To do so, we are using the R `t()` function.\n", "\n", "&#x1F4DD; **Activity**: Set the values for these two arguments before running the cell below i.e. replace ... with either `TRUE` or `FALSE`."]}, {"cell_type": "code", "execution_count": null, "id": "objective-struggle", "metadata": {}, "outputs": [], "source": ["pca_results = prcomp(t(df_nonzero_var), scale=..., center = ...)"]}, {"cell_type": "markdown", "id": "announced-memorabilia", "metadata": {}, "source": ["We can use R `summary()` function to get a summary of the PCA result.\n", "Let's inspect the result by running the following cell.\n", "\n", "&#x1F4DD; **Activity**: \n", "- How many PCs are resulted from the PCA analysis?\n", "- Is there a relation between the number of PCs and the number of samples?\n", "- What percentage of the variance in the data is explained by the first two PCs? "]}, {"cell_type": "code", "execution_count": null, "id": "impossible-principal", "metadata": {}, "outputs": [], "source": ["summary(pca_results)"]}, {"cell_type": "markdown", "id": "fabulous-channels", "metadata": {}, "source": ["We can plot a 2-D plot using the first `2` PCs using the following cell. For visualization, we are using `ggplot` package. In this exercise, we skip the visualization details.\n", "\n", "&#x1F4DD; **Activity**:\n", "- Are cancer samples separated from normal samples?\n", "- If yes, along which PC, this separation is visible?\n", "- Are normal samples, more similar to each other or cancer samples?"]}, {"cell_type": "code", "execution_count": null, "id": "circular-quest", "metadata": {}, "outputs": [], "source": ["library(ggplot2)\n", "\n", "\n", "## Creating a data frame out of PCA analysis result\n", "pca_results_df = data.frame(Sample = rownames(pca_results$x),\n", "                             X = pca_results$x[,1],\n", "                             Y = pca_results$x[,2])\n", "\n", "## Calculating the variance explained by each principal components\n", "pca_var = pca_results$sdev^2\n", "pca_var_percent = round(pca_var/sum(pca_var)*100, 1)\n", "\n", "## Plotting\n", "ggplot(data = pca_results_df, aes(x = X, y = Y, label = Sample)) +\n", "geom_point() + \n", "scale_x_continuous(limits = c(-125, 125)) + \n", "scale_y_continuous(limits = c(-125, 125)) +\n", "geom_text(mapping = aes(x = X, y = Y+10, label = Sample), size = 5) +\n", "  xlab(paste(\"PC1: \", pca_var_percent[1], \"%\")) + \n", "  ylab(paste(\"PC2: \", pca_var_percent[2], \"%\")) + \n", "  ggtitle(\"PCA analysis result\") +\n", "  theme_bw()"]}, {"cell_type": "markdown", "id": "nominated-exchange", "metadata": {}, "source": [" "]}, {"cell_type": "markdown", "id": "incorporate-blackberry", "metadata": {}, "source": [" "]}, {"cell_type": "markdown", "id": "lyric-holly", "metadata": {}, "source": ["Another way to inspect, analyze our high-dimensional data, is via clustering the genes and/or samples. For example, clustering genes groups together genes with similar expression profile across samples. There are 3 major classes of clustering methods: 1) hierarchical clustering, 2) partitioning, and 3) density-based methods. Here, we apply hierarchical clustering to our data. Often, the hierarchical results is followed by visualization using heat maps as we will see below."]}, {"cell_type": "markdown", "id": "forward-campaign", "metadata": {}, "source": ["## Hierarchical clustering"]}, {"cell_type": "markdown", "id": "twenty-diagnosis", "metadata": {}, "source": ["To perform hierarchical clustering, we need to have a measure that tells the distance between two samples/genes (the smaller the distance, the more similar the samples/genes are; this is referred to as *distance metric*). Once we have a way to calculate the distance, we start by grouping the samples/genes with the least distance to each other and continue until all samples/genes are grouped. In this approach, sometimes we need to find the distance of a sample/gene to a groups of samples/genes. There are few ways to find such distances (this is referred to as *linkage method*). For example, to calculate the distance between a sample *A* and a group/cluster of 2 samples (*B* and *C*) using the \"complete\" linkage method, we use the maximum distance of *A* and *B* or *A* and *C* to be the distance between *A* and the cluster that contain *B* and *C*.\n", "\n", "There are several distance metrics available e.g. Euclidean distance, Manhattan distance, and correlation distance to name a few. The choice of the proper distance metric is important. For example, it has been known that Euclidean distance is sensitive to scaling and differences in average expression level whereas correlation is more robust to these changes. In this example, we use **correlation distance** for **distance metric** and **complete** linkage for **linkage method**.\n", "\n", "We use R `hclust()` function to perform hierarchical clustering (use `?hclust` to familiarize yourself with the function). using the R `plot()` function, we can visualize the result which is a *cluster dendrogram* which shows the clusters on the X-axis as well as the distance between the clusters on the Y-axis."]}, {"cell_type": "code", "execution_count": null, "id": "based-language", "metadata": {}, "outputs": [], "source": ["## defining a function that calculates the correlation distance\n", "correlation_dist = function(x, method = \"pearson\"){\n", "    corr_distance = as.dist((1-cor(x, method=method))/2) \n", "    return(corr_distance)\n", "}\n", "\n", "## Hierarchical clustering\n", "hc = hclust(correlation_dist(df_nonzero_var), method = \"complete\")\n", "\n", "## plotting the hierarchical clustering results\n", "plot(hc)"]}, {"cell_type": "markdown", "id": "hairy-schema", "metadata": {}, "source": ["&#x1F4DD; **Activity**:\n", "- Look at the Y-axis of the plot above! What is the approximate distance between `GSM1778906` and `GSM1778907`? \n", "- Look at the Y-axis of the plot above! What is the approximate level of Pearson correlation between `GSM1778906` and `GSM1778907`?\n", "- Compare PCA plot with the hierarchical clustering plot. Are the two closest normal samples in the PCA differ from the two closest normal samples in the hierarchical clustering? What could be the reason?\n", "- In the following cell, change the distance metric to Euclidean distance (**hints**: you need to use the R `dist()` function. Also, you need to transpose the data). \n", "    - Do you see any difference between this and earlier plot? Would scaling the data change the final result? "]}, {"cell_type": "code", "execution_count": null, "id": "sustained-exchange", "metadata": {}, "outputs": [], "source": ["## Do the activity in this cell"]}, {"cell_type": "markdown", "id": "meaningful-linux", "metadata": {}, "source": ["### Heat map"]}, {"cell_type": "markdown", "id": "attractive-vegetable", "metadata": {}, "source": ["As mentioned above, we can use heat map to visualize our data after the hierarchical clustering. Here, we use `pheatmap` R package which provides us with `pheatmap()` function to perform both hierarchical clustering as well as visualization.\n", "\n", "Below, we are using the top 1000 most variable genes for the visualization.\n", "\n", "&#x1F4DD; **Activity**:\n", "- Name a couple of reasons why we only use a fraction of data for this purpose?\n", "- Compare the clustering of the samples with earlier results. Do you see any difference?\n", "- Look at the heat map, are there groups of genes that behave differently across different sample types (i.e. normal and tumor)?\n", "- What is the purpose of setting `scale = 'row'` (**hint**: ?pheatmap)?\n", "- What is the purpose of setting `show_rownames=FALSE` (**hint**: ?pheatmap)? "]}, {"cell_type": "code", "execution_count": null, "id": "passive-sacrifice", "metadata": {"scrolled": false}, "outputs": [], "source": ["library(pheatmap)\n", "\n", "df_vars_order = order(df_vars, decreasing = TRUE) # returns the indices of rows when order in a decreasing order\n", "df_ordered = df[df_vars_order,] # Using the indices from above to reorder our data frame; Remember that df \n", "                                # contain also rows with variance 0, but after reordering, these row are all\n", "                                # at the bottom.\n", "\n", "pheatmap(as.matrix(df_ordered[1:1000,]), \n", "         scale = 'row',\n", "         clustering_distance_rows = 'correlation', # Tells to use the Pearson correlation distance for clustering\n", "         clustering_distance_cols = 'correlation',\n", "         show_rownames = FALSE,  \n", "         main = 'Heat map of top variable genes',\n", "        )"]}, {"cell_type": "markdown", "id": "occasional-conspiracy", "metadata": {}, "source": [" "]}, {"cell_type": "markdown", "id": "similar-disclosure", "metadata": {}, "source": [" "]}, {"cell_type": "markdown", "id": "identified-plasma", "metadata": {}, "source": ["## Differential expression analysis using `DESeq2`"]}, {"cell_type": "markdown", "id": "academic-viewer", "metadata": {}, "source": ["Based on the heat map, we see that there genes that are more expressed in normal samples than in tumor sample or vice versa. One might hypothesize that these genes have something to do with the development of cancer or the absence of it. So, how can we reliably find such genes that are differentially expressed (DE) across different sample types? \n", "\n", "At its simplest, we can go through genes, one gene at a time, and use an appropriate statistical test that tells whether the two groups are different from each other in terms of their gene expression (our null hypothesis is that the groups are similar or data from both groups come from one distribution). However, we have to take into consideration several possible issues. \n", "\n", "For example, to be able to compare the groups, they need to be comparable. Thus, we need a way to normalize the data before the comparison. But what is a proper normalization approach for our data? \n", "\n", "For another example, consider answering the question of what is the appropriate statistical test for our purpose knowing that each statistical test makes assumptions about the underlying data distribution. What is the distribution of our data? \n", "\n", "Let's consider another example: as we have thousands of genes to test, we end up in a multiple hypothesis testing situation. If we do not correct for this, in our analysis result, we might have some genes labeled as DE where in reality they are not (**note**: by removing the genes with variance `0`, we somehow alleviate this problem because in total, we do less test but the problem still exists). \n", "\n", "Luckily, there has been several statistical methods developed for finding DE genes that try to address many of the issues relevant to the differential expression analysis (e.g. [DESeq2](http://bioconductor.org/packages/release/bioc/html/DESeq2.html), [limma](http://bioconductor.org/packages/release/bioc/html/limma.html), and [edgeR](http://bioconductor.org/packages/release/bioc/html/edgeR.html) ). \n", "\n", "Here, we use a popular method called `DESeq2`. `DESeq2` starts by using the count data and performs all required normalization, modeling, testing, and multiple hypothesis testing correction for us.  "]}, {"cell_type": "code", "execution_count": null, "id": "ambient-basic", "metadata": {}, "outputs": [], "source": ["## loading DESeq2\n", "library(DESeq2)\n", "\n", "## Loading the meta data for our samples\n", "coldata = read.csv('https://raw.githubusercontent.com/eafyounian/2021_summer_school_bioinf/main/resources/GSE70466-metadata.txt', \n", "                    sep = '\\t', \n", "                    row.names = 'Sample_geo_accession',\n", "                   )\n", "coldata = coldata['cell.line']\n", "\n", "## creating DESeqDataSet\n", "dds = DESeqDataSetFromMatrix(countData = as.matrix(df_nonzero_var), \n", "                              colData = coldata, \n", "                              design = ~cell.line\n", "                             )\n", "\n", "## Setting the cell line against which the comparison is performed \n", "dds$cell.line <- relevel(dds$cell.line, ref = \"PrEC\")\n", "\n", "## Perform the differential expression analysis\n", "dds = DESeq(dds)\n", "\n", "## Correcting for the multiple hypothesis issue\n", "res = results(dds, alpha = 0.0001)\n", "\n", "## Keeping only genes that their adjusted p-value is less than 0.0001.\n", "## This threshold sets the FDR to 0.01% i.e. 0.01% of our DE genes are false positives\n", "res_subset = subset(res, padj < 0.0001)\n", "\n", "## Ordering the DE genes based on the log2 fold change\n", "res_subset_ordered = res_subset[order(res_subset$log2FoldChange),]\n", "\n", "## Write the results into file(s) for future use\n", "write.table(as.data.frame(res_subset_ordered), \n", "            file = 'LNCaP_vs_PrEC_DE_genes.txt', \n", "            sep = '\\t', \n", "            quote=FALSE\n", "           )\n", "\n", "write.table(rownames(res_subset_ordered), \n", "            file = 'LNCaP_vs_PrEC_DE_gene_names.txt', \n", "            sep = '\\t', \n", "            quote=FALSE, \n", "            row.names=FALSE\n", "           )\n", "\n", "\n", "## Looking at the summary of the data\n", "summary(res_subset_ordered)"]}, {"cell_type": "markdown", "id": "fluid-protocol", "metadata": {}, "source": ["&#x1F4DD; **Activity**:\n", "- Use the `head(res_subset_ordered, 1)` to see the first DE gene. \n", "    - What is the name of this gene?\n", "    - In which sample type is this gene more expressed (up-regulated)?\n", "- Use the `tail(res_subset_ordered, 1)` to see the last DE gene. \n", "    - What is the name of this gene?\n", "    - In which sample type is this gene more expressed?\n", "- Inspect the summary results in the above cell.\n", "    - How many genes are DE?\n", "        - OF these DE, around how many do we expect to be false positive?\n", "    - How many genes are up-regulated in LNCaP?\n", "    - How many genes are down-regulated in LNCaP?"]}, {"cell_type": "code", "execution_count": null, "id": "turkish-saint", "metadata": {}, "outputs": [], "source": ["## Do the activity in this cell"]}, {"cell_type": "code", "execution_count": null, "id": "generic-worse", "metadata": {}, "outputs": [], "source": ["## Do the activity in this cell"]}, {"cell_type": "markdown", "id": "willing-helicopter", "metadata": {}, "source": ["Let's plot the heat map of only the DE genes similar to the heat map we created above.\n", "\n", "&#x1F4DD; **Activity**:\n", "- Compare this heat map with the earlier heat map. Name some of the changes that you notice?"]}, {"cell_type": "code", "execution_count": null, "id": "demonstrated-privilege", "metadata": {}, "outputs": [], "source": ["## Subsetting our original data frame by keeping only DE genes\n", "DE_gene_expr = df[rownames(res_subset),]\n", "\n", "pheatmap(as.matrix(DE_gene_expr), \n", "         scale = 'row',\n", "         clustering_distance_rows = 'correlation', # Tells to use the Pearson correlation distance for clustering\n", "         clustering_distance_cols = 'correlation',\n", "         show_rownames = FALSE,  \n", "         main = 'Heat map differentially expressed genes',\n", "        )"]}, {"cell_type": "markdown", "id": "thorough-greek", "metadata": {}, "source": ["**Note**: Here, we just scratched the surface using `DESeq2`. For more more information, please read [this](http://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) tutorial written by the creators of the tool."]}, {"cell_type": "markdown", "id": "documentary-ultimate", "metadata": {}, "source": ["So far so good! We now have a list of genes that are differentially expressed across the two sample types. How are we going to make sense of this list? One of the simplest ways is to go through the list, one gene at a time, and see if we can spot a gene that sounds interesting with respect to the disease we are studying. But we have thousands of DE genes in this experiment. One way to approach this issue here is via enrichment analysis. The basic idea is to see whether e.g. a biological pathway is enriched with DE genes. If we find such a pathway, we could hypothesize that this pathway may be related to the condition under study. "]}, {"cell_type": "markdown", "id": "cellular-combat", "metadata": {}, "source": ["##  Functional and enrichment analysis"]}, {"cell_type": "markdown", "id": "answering-lighter", "metadata": {}, "source": ["Here, we perform functional and enrichment analysis over all differentially expressed genes using an [online service](https://david.ncifcrf.gov/home.jsp) called Database for Annotation, Visualization and Integrated Discovery (DAVID).\n", "\n", "When we performed differential expression analysis, we saved the results in a file called: `LNCaP_vs_PrEC_DE_gene_names.txt`. Following the steps below, we will check if we can find any enriched pathways.\n", "\n", "- Go to the GenePattern Notebook Home Page tab in your browser. You should be able to see a file named: `LNCaP_vs_PrEC_DE_gene_names.txt`. Download this device to your local machine by selecting the check box next to the file and pressing the <kbd>Download</kbd> button that appears. We will upload this file into DAVID. \n", "- Open a new browser tab by following [this](https://david.ncifcrf.gov/home.jsp) link.\n", "- On the web-page, click on the Start Analysis. \n", "- On the left panel make sure that upload tab is selected. \n", "    - In Step 1: `Enter Gene List`, select <kbd>Choose File</kbd> and upload the file you downloaded to your local machine. \n", "    - In Step 2: `Select Identifier` select: `OFFICIAL_GENE_SYMBOL`.\n", "    - In Step 2a: `Select species` select `Homo sapiens` by typing it in the box.\n", "    - In Step 3: `List Type`, select `Gene List`. \n", "    - In Step 4, press the <kbd>Submit List</kbd> button.\n", "- Now, you should be able to see the message: `Step 1. Successfully submitted gene list`.\n", "- Under `Step 2. Analyze above gene list with one of DAVID tools`, press the `Functional Annotation Tool` link.\n", "- Under `Annotation Summary Results`, select `Pathways`.\n", "- Finally, press the <kbd>Chart</kbd> button next to `KEGG_PATHWAY`. Inspect the results in the window that opens up. The results are already sorted based on the Benjamini-adjusted p-values.\n", "\n", "&#x1F4DD; **Activity**:\n", "- Do you spot any interesting enriched pathways?\n", "- Under the term column, press the link to one enriched pathway that interests you. This should open up a pathway chart with DE genes marked with blinking starts on the pathway. "]}, {"cell_type": "markdown", "id": "basic-gender", "metadata": {}, "source": ["## Conclusion"]}, {"cell_type": "markdown", "id": "parliamentary-combine", "metadata": {}, "source": ["Here, we conclude this very short analysis of RNA-seq data. Here, we scratched only the surface. We did not go very deep into the details of the analyses we did, and we left out many other possible analyses that could have been done. "]}, {"cell_type": "markdown", "id": "beginning-relay", "metadata": {}, "source": ["### Further reading"]}, {"cell_type": "markdown", "id": "nominated-pantyhose", "metadata": {}, "source": ["Here are some resources that might be helpful for people interested in further learning about the RNA-seq analysis.\n", "- [A survey of best practices for RNA-seq data analysis](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0881-8)\n", "- [RNA-seqlopedia](https://rnaseq.uoregon.edu/)\n", "- [DESeq2 tutorial](http://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)\n", "- [Informatics for RNA-seq: A web resource for analysis on the cloud](https://github.com/griffithlab/rnaseq_tutorial)\n", "- [Pathway Commons guid on RNA Sequencing Analysis](https://www.pathwaycommons.org/guide/primers/data_analysis/rna_sequencing_analysis/)"]}, {"cell_type": "markdown", "id": "twenty-serve", "metadata": {}, "source": ["Thank you for your attention and good luck! &#x1F389;"]}], "metadata": {"genepattern": {"repository_url": "http://notebook.genepattern.org/services/sharing/notebooks/414/"}, "kernelspec": {"display_name": "R 3.6", "language": "R", "name": "r3.6"}, "language_info": {"codemirror_mode": "r", "file_extension": ".r", "mimetype": "text/x-r-source", "name": "R", "pygments_lexer": "r", "version": "3.6.3"}, "toc": {"base_numbering": 1, "nav_menu": {}, "number_sections": false, "sideBar": true, "skip_h1_title": false, "title_cell": "Table of Contents", "title_sidebar": "Contents", "toc_cell": false, "toc_position": {}, "toc_section_display": true, "toc_window_display": false}}, "nbformat": 4, "nbformat_minor": 5}